### ISSUE TITLE
Resource not re-sent if clients unsubscribe and then resubscribe

### ISSUE THREAD

**Issue by menghanl** [envoyproxy/go-control-plane#431]
The scenario:

Client is watch for resource A
```
req(A, "")
resp(A, v1)
req(A, v1) // this is ACK
```

user unsubscribe for A
```
req([], v1) // this removes A from the requested resources
```

user resubscribe for A
```
req(A, v1)
```

It's expected that the control plane will resend resource for A to the client, even though there's no update in the resource, because of the unsubscribe and resubscribe.

But go-control-plane doesn't re-send the resource.

**Comment by menghanl** [comment]
The spec doesn't talk about this scenario explicitly.

In the [Unsubscribing From Resources](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#unsubscribing-from-resources) section
> unsubscribing to a set of resources is done by sending a new request containing all resource names that are still being subscribed to but not containing the resource names being unsubscribed to

In the [Resource updates](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#resource-updates) section
> Envoy may update the list of resource_names it presents to the management server in each DiscoveryRequest that ACK/NACKs a specific DiscoveryResponse. In addition, Envoy may later issue additional DiscoveryRequests at a given version_info to update the management server with new resource hints

**Comment by menghanl** [comment]
cc @markdroth @dfawley

**Comment by snowp** [comment]
This seems like a bug to me. I assume this happens because the code right now assumes that if the requested version matches the server version then nothing needs to be done, but it sounds like we need to track subscriptions even in SotW to re-push resources for these kind of subscription changes.

@htuch can you confirm that what this issue talks about is the expected behavior?

@alecholmez We might need some kind of subscription state for SotW as well based on this issue, maybe there are options for consolidating the new delta state with sotw.

**Comment by markdroth** [comment]
Yes, I think this is a bug.  The xDS server definitely needs to explicitly track the set of resources that the client is subscribed to.

Another relevant part of the spec is [Knowing When a Requested Resource Does Not Exist](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#knowing-when-a-requested-resource-does-not-exist):

> Note that even if a requested resource does not exist at the moment when the client requests it, that resource could be created at any time. Management servers must remember the set of resources being requested by the client, and if one of those resources springs into existence later, the server must send an update to the client informing it of the new resource. Clients that initially see a resource that does not exist must be prepared for the resource to be created at any time.

The way I think of an xDS server, its job is to basically match up the states of a database containing a set of resources with the states of the individual client streams.  Each client stream has an associated list of resource names that it is subscribed to, along with the corresponding version of each resource that has already been sent to the client.  Whenever the client changes the set of resources it is subscribing to, new resources may need to be sent to the client.  Whenever a resource in the database gets updated (where creating and removing a resource are just special cases), that change needs to be propagated to any streams that are subscribing to that resource.  (In this sense, the version tracking in SotW is IMHO actually harder to implement than the state tracking for the incremental protocol, because the semantics around the resource type instance version are actually much harder to understand.)

Note that in general, the server cannot rely on the version sent by the client to decide what set of resources it needs to send; it should always pay attention to changes in the set of resources that the client is subscribing to in its requests.  The only case where the server may not send a resource requested by the client is upon stream reconnection, and even then this optimization works only in a very narrow set of circumstances, as per the [ACK/NACK and resource type instance version](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#ack-nack-and-resource-type-instance-version) section of the spec:

> Note that the version for a resource type is not a property of an individual xDS stream but rather a property of the resources themselves. If the stream becomes broken and the client creates a new stream, the client’s initial request on the new stream should indicate the most recent version seen by the client on the previous stream. Servers may decide to optimize by not resending resources that the client had already seen on the previous stream, but only if they know that the client is not subscribing to a new resource that it was not previously subscribed to. For example, it is generally safe for servers to do this optimization for wildcard LDS and CDS requests, and it is safe to do in environments where the clients will always subscribe to exactly the same set of resources.

**Comment by alecholmez** [comment]
@snowp I think we can certainly abstract a pattern out of the delta logic that we could apply within SOTW.

Luckily I defined the `StreamState` as a global server package so we could internally reference that pattern throughout the code

**Comment by alecholmez** [comment]
This is a big bug and I think I will go ahead and take this

**Comment by steeling** [comment]
any update on this?

**Comment by tony612** [comment]
@alecholmez Any plan to fix this?

**Comment by bayoumymac** [comment]
@alecholmez any update on this?

**Comment by duxin40** [comment]
@alecholmez any update on this?

**Comment by kkalin68** [comment]
I know there is a PR that will address the issue. Meanwhile I made a patch for `v0.12.0` that seems to solve the issue on current codebase

```
diff --git a/pkg/cache/v3/simple.go b/pkg/cache/v3/simple.go
index ebf63f5b6..6dbcf6398 100644
--- a/pkg/cache/v3/simple.go
+++ b/pkg/cache/v3/simple.go
@@ -411,11 +411,19 @@ func (cache *snapshotCache) CreateWatch(request *Request, streamState stream.Str
 	if exists {
 		knownResourceNames := streamState.GetKnownResourceNames(request.GetTypeUrl())
 		diff := []string{}
+		subscribed := make(map[string]bool)
 		for _, r := range request.GetResourceNames() {
+			subscribed[r] = true
 			if _, ok := knownResourceNames[r]; !ok {
 				diff = append(diff, r)
 			}
 		}
+		for k := range knownResourceNames {
+			if _, ok := subscribed[k]; !ok {
+				delete(knownResourceNames, k)
+			}
+		}
+		streamState.SetKnownResourceNames(request.GetTypeUrl(), knownResourceNames)
 
 		cache.log.Debugf("nodeID %q requested %s%v and known %v. Diff %v", nodeID,
 			request.GetTypeUrl(), request.GetResourceNames(), knownResourceNames, diff)
```

**Comment by carsonoid** [comment]
> for k := range knownResourceNames {
> +			if _, ok := subscribed[k]; !ok {
> +				delete(knownResourceNames, k)
> +			}
> +		}
> +		streamState.SetKnownResourceNames(request.GetTypeUrl(), knownResourceNames)

I can confirm. I am able to reliably recreate this issue using go grpc xds clients and applying this patch fixes it! I'm wondering if there a downside to using this fix over the fixes and changes proposed PRs attached to this issue. This seems like a much smaller and more focused fix.

---

For those who are curious, it's actually fairly easy to recreate this locally:

* Create two grpc clients in one process
* One uses `ResourceA` and has a very short idle timeout (I used 250ms)
* Two uses `ResourceB` and has a normal timeout

Then have each client use do a grpc call every second, the first client will basically become blocked instantly and never work. It will simply always report `Received error from the name resolver: produced zero addresses`. The broken client will then stay in this state until an xds update happens and the watches are recreated.

**Comment by valerian-roche** [comment]
This [PR](https://github.com/DataDog/go-control-plane/pull/10) in our fork has been implemented to address this issue. We are currently validating the impact on grpc and envoy clients. 
Other issues are also being addressed to properly handle grpc clients in those [PRs](https://github.com/DataDog/go-control-plane/pulls?q=is%3Apr+is%3Aclosed+label%3Ato-upstream). There is also a [code fix in upstream grpc](https://github.com/grpc/grpc-go/pull/7026) (for go) to address an improper wildcard subscription which will be released in grpc-go 1.63.0

**Comment by easwars** [comment]
gRPC-Go uses the go-control-plane as the management server for our tests and we are seeing multiple tests flaking because of this bug. Can a fix for this be prioritized, please. Thanks.

**Comment by easwars** [comment]
I've just sent out a PR that uses a patch mentioned in https://github.com/envoyproxy/go-control-plane/issues/431#issuecomment-2021010775. I also added a unit test for the same scenario and was able to verify that gRPC-Go tests that exercise the unsubscribe/resubscribe functionality work after this change.
